pkg/controller/backup_controller.go

```diff
@@ -587,19 +624,27 @@ func recordBackupMetrics(log logrus.FieldLogger, backup *velerov1api.Backup, bac
 	serverMetrics.RegisterVolumeSnapshotFailures(backupScheduleName, backup.Status.VolumeSnapshotsAttempted-backup.Status.VolumeSnapshotsCompleted)
 }
 
-func persistBackup(backup *pkgbackup.Request, backupContents, backupLog *os.File, backupStore persistence.BackupStore, log logrus.FieldLogger) []error {
+func persistBackup(backup *pkgbackup.Request,
+	backupContents, backupLog *os.File,
+	backupStore persistence.BackupStore,
+	log logrus.FieldLogger,
+	csiVolumeSnapshots []*snapshotv1beta1api.VolumeSnapshot,
+	volumeSnapshotContents []*snapshotv1beta1api.VolumeSnapshotContent,
+) []error {
+
 	errs := []error{}
 	backupJSON := new(bytes.Buffer)
 
 	if err := encode.EncodeTo(backup.Backup, "json", backupJSON); err != nil {
 		errs = append(errs, errors.Wrap(err, "error encoding backup"))
 	}
 
-	volumeSnapshots := new(bytes.Buffer)
-	gzw := gzip.NewWriter(volumeSnapshots)
+	// Velero-native volume snapshots
```

Comment:
There's a lot of duplication in this function for writing resources to a gzipped JSON writer here (backup.VolumeSnapshots, backup.PodVolumeBackups, csiVolumeSnapshots, volumeSnapshotContents, backup.BackupResourceList()) - we can probably extract a helper func, and/or change this code to be a loop over the relevant items, where the loop body handles the json+gzip writing.

Reply:
How important is de-duping the internals of this function to this PR? I can see the need, and can spend some time on it, just not sure of where to include it.

Reply:
Ok, latest commits have moved the serialization/compression into a helper function, so hopefully this is easier to follow.